doc/api-management.md

```diff
@@ -0,0 +1,59 @@
+# API Management
+
+> Guidance below for Theia framework. End products will get better development experience by using private visibility.
+
+Theia is a framework embracing openness, extensibility, and customizability as much as possible:
+- API defaults to public for clients.
+- API defaults to protected for extenders.
+- Language constructions prohibiting runtime access to internals are never used and so on.
+
+Usually, the version management is built around API visibility.
+Particularly, if a public API is broken then the major release is required.
+
+Following the conventional approach is not practicable for Theia since all APIs are more or less public.
+It will slow down API innovation, accumulate technical debt or require many major releases.
+
+Instead, we require the major release only if a stable API is broken.
+Breaking an experimental API can be done with the minor release,
+but not necessary if such API does not have sufficient adoption.
+
+## Stability
+
+Conceptually, an API is all assumptions adopters depend on to make use or extend Theia.
+A stable API is based on assumptions that are not subject to change.
+It does not matter whether an API is public or internal, or whether it is used a lot or never.
+
+> For instance, the language server protocol depends on such data types like URIs and positions
+since they don't change from language to language. It ensures its stability.
+
+The API stability indicated by adding, exclusively, `@experimental` or `@stable` js-doc tags.
+An API without a stability tag considered to be experimental.
+
+### Experimental
+
+- All new APIs should always be added as `experimental`
+since it's almost impossible to get stable API right first time.
+- Usually, experimental APIs don't require the stability tag,
+but if a new member is added to `stable` API then it should be explicitly annotated.
+- Such APIs should not have extensive documentation to avoid the maintenance burden.
+- They can be changed or removed without the deprecation cycle if it was not widely adopted.
+- Adoption should be measured by the number of internal clients or based on the feedback of Theia contributors and committers.
+
+### Stable
+
+- Experimental APIs can be graduated to `stable` after at least one major release.
```

Comment:
Just to confirm: This means that a newly added experimental API cannot be made stable until we step the major version? e.g. a new API in v1.x can only be made stable starting at v2.0.0?

Reply:
Yes, do you know how it could be phrased better?

The major version will mean that we either:
- have experimental APIs with sufficient adoption and not likely to change, so they are matured to be stable
- or we broke stable APIs. Ideally it should never happen if we graduate APIs properly.

Reply:
Why this limit? With Semver adding new API in minor versions is usually ok as long as we don't break existing, no?

Reply:
> Why this limit? With Semver adding new API in minor versions is usually ok as long as we don't break existing, no?

The approach is to work around semver, i.e. we will follow it only for stable APIs. It won't restrict us to break new APIs without major version.

If someone wants to enforce some API to be stable and follow semver, they will need to ask for the major release with GH issue to annotate and document such APIs. I think it is reasonable since it is extra effort on our side:
- to analyze, stabilize and documents APIs
- to maintain their backward compatibility afterwards

We could decide then for a new major release if we have enough of such API requests.

I could also change it to something like `at least 3 minor releases` instead of major. Or maybe simply that there should be explicit API request via a GH issue.

Reply:
We have to consider how the frequency of major releases affect adopters: While we would embrace stabilizing API (often major releases) each of these major release also sends a signal that API has bean broken.

I'd like the "3 minor releases" much better in this regard

Reply:
ok, it is true that many have already assumptions about major release semantic, so i will remove it and replace with a requirement for explicit stable API request as GH issue, it also aligns well with https://github.com/eclipse-theia/theia/pull/7424#discussion_r398389145 and https://github.com/eclipse-theia/theia/pull/7424#discussion_r398390178

Timing wise it will be decided from case to case.